### DISCUSSION TITLE
Slack notifications for deployment

### DISCUSSION THREAD

**Discussion by marko-hologram** [coollabsio/coolify#2210] [Feature Requests]
Hey, first of all thank you again for making Coolify. Love using it so far!

Unfortunately, as this software is broad in scope, there are inevitably some things that are missing for some of us using it.

I do see that there are already Notifications as a feature, which is nice, but I think Email, Telegram and Discord are kinda limiting here. Many companies use Slack as a hub for all messaging and these types of notifications are no different. 

Many other tools already have Slack integration because it's such a popular messaging app and many teams use these integrations.

### Opt-in by default

By default, Notifications are disabled and are opt-in. This makes sense because they do require integration with Slack API and some setup.

### Per resource notifications

I work in an agency and we work on many projects in parallel. Global notifications that go into a single channel are too much spam and people usually mute these type of channels which then makes them useless because nobody is looking at them.

What I would ideally prefer is **per resource** notifications. I specifically highlight the per resource part because I think that's important. Per project notifications would be ok, but still not ideal. Let's say that in a project's staging environment I have 3 resources deployed, 1 database, 1 backend service and 1 frontend application. I only care when backend and frontend are deployed. In case database gets updated to latest version or re-deployed for some reason, that's not really something the rest of the team needs to be spammed about.

I would also prefer per resource notifications because you might just want to send notifications for "production" environment and not "staging" environment.

### Sending notifications to any Slack channel

Current Notifications in Coolify work fine if you are a solo developer or a single team, but in a multi team environment they are not enough. Having global Notification settings that assume all notifications go into a single channel of communication is limiting and results in spam and in notifications being ignored as I mentioned above.

Ideally, notifications could be sent to any channel in a specific Slack workspace by specifying the ID of the channel. ID would be preferred to channel name because ID is unique and cannot be updated, unlike channel name. ID is more resilient.

If you work in an agency like I do, this is crucial because it means that messages can be delivered into channels that are actually related to that project. Just dumping everything into a single channel is horrible and messy and it also means people that work on a single project will get spammed with notifications from X other projects.

### Custom notification message

Ideally, each enabled notification message can be edited to send a custom message for that specific resource. It would be great if messages can be templated in some way to make them dynamic.

For example, something like this:

```
Deploy successful!

Project: {{ project_name }}
Resource: {{ resource_name }}
Branch: {{ vcs_branch }}
URL: {{ deployment_url }}
Logs: {{ resource_logs_url }}
```

Would send a message like this:

![image](https://github.com/coollabsio/coolify/assets/22904204/df0886a8-0f79-4fc9-b3d7-df7c580a9789)

With a custom template message like this, same message can be used across the board if a team wants to have a standard message shape. Values in brackets (`{{ }}`) would get replaced with their actual values for that deployment.

Things I would love to have as variables in these messages are:

- basic project information such as name
- basic resource information such as name
- name of the branch that was used for the deployment, not SHA of the commit being deployed, but actual branch name (e.g. "main")
  - I used "vcs_" here as a generic prefix for all "Version Control Systems", but if your plan is to only support git, then it can be renamed to "git_branch"
- URL where the resource was deployed
  - in case a resource has multiple URLs pointed to it, maybe just pick the first one?
- links to specific parts of that deployment such as logs or just configuration page

### Color coded messages

For example, Jenkins Slack integration adds simple color coding to messages which can quickly tell you if something is ok or it failed.

Here is an example:

![image](https://github.com/coollabsio/coolify/assets/22904204/a55619be-dda0-4061-ad2d-98c705c037ee)

Content of these messages is cut off because I cannot show that information and it's irrelevant really for this.

### Preview deployments

Ideally, notifications for these can be enabled/disabled on a per resource basis. If a project has many PRs daily, these can turn into spam. They don't necessarily need to be sent to a channel to notify everyone, including non developer people present in the channel. Staging/Production environment deployments are more valuable to the broader audience like product managers, designers, QA and sales.

## Existing integrations / Inspiration

I think Jenkins does this Slack integration right (https://plugins.jenkins.io/slack/) and I've based most of my ideas on that. 

It allows you to specify global channel ID where messages go by default, but it also allows per project override of these options. You can also send messages on successful deploy, failed deploy, unstable deploy etc. based on status values that Jenkins has for deploys. For Coolify this would most likely be just Success and Fail/Error I guess?

## How can I help?

And lastly, how can I help with this? I didn't just want to randomly start poking into Coolify because this is a decent chunk of work and I saw you prefer suggestions/proposals before any serious work is being done.

If I can somehow help with this, please let me know.

**Comment by FaureAlexis** [comment]
Agree, if it's okay @andrasbacsai I can create a PR with this feature within a week

**Comment by japurillo** [comment]
This would be a really great feature. Telegram and Email doesnt really cut it.

**Comment by swizzmagik** [comment]
any progress here?

**Comment by isaackoz** [comment]
In your project under General>Pre/Post Deployment Commands, just add the following command:
`curl -X POST -H 'Content-type: application/json' --data '{"text":"Deployment was successful."}' YOUR_SLACK_WEBHOOK_URL`

**Reply by AprilNEA** [reply]
how to capture failure?

**Reply by marko-hologram** [reply]
This is a possibility, but it's not really a replacement for proper integration of this feature.

**Reply by isaackoz** [reply]
> This is a possibility, but it's not really a replacement for proper integration of this feature.

I agree, but for a quick and dirty implementation this worked decent enough. Although it only works when the build is successful. 

Would be interesting if there were a post deployment hook that ran regardless of success/failure and instead passed in variables to the script so you could do something like `STATUS=success sh -c 'python3 /usr/local/bin/my_script.py'`. This way you could integrate with anything you'd like.

I'm still getting to know Coolify so if this already exists or is possible another way, please let me know 😄

**Reply by etgohomeok** [reply]
I guess you could also have a pre-deployment hook that indicates that the deployment is starting and if you don't get the post-deployment hook in a reasonable amount of time after that, you can assume it failed?

**Comment by etgohomeok** [comment]
Just chiming in that I think the best solution would just be to integrate Apprise into the notifications (or at least the ability to send hooks to our own Apprise API) since that would enable all kinds of notifications, including but not limited to Slack. No need to reinvent the wheel.

**Reply by marko-hologram** [reply]
To be honest, whatever works as long as I can get the features I described above.

**Comment by straatrakker** [comment]
Would love to see this feature.

**Comment by sbushell92** [comment]
Also would love to see slack integration for successful/failed deployments.

**Comment by marvinvr** [comment]
Hey, I implemented this as part of #4264 Would be great to get this reviewed.
